run-tests.php

```diff
@@ -1349,73 +1348,37 @@ function run_all_tests_parallel($test_files, $env, $redir_tested) {
 	echo "====⚡️==== WELCOME TO THE FUTURE: run-tests PARALLEL EDITION ====⚡️====\n";
 	echo "====⚡️===========================================================⚡️====\n";
 
-	// Because some of the PHP test suite has not been written with
-	// parallel execution in mind, it is not safe to just run any two tests
-	// concurrently.
-	// Therefore, we divide the test set into directories and test multiple
-	// directories at once, but not multiple tests within them.
-
-	$testDirsToGo = [];
-
+	// Each test may specify a list of conflict keys. While a test that conflicts with
+	// key K is running, no other test that conflicts with K may run. Conflict keys are
+	// specified either in the --CONFLICTS-- section, or CONFLICTS file inside a directory.
+	$dirConflictsWith = [];
+	$fileConflictsWith = [];
 	foreach ($test_files as $file) {
-		$dirSeparator = strrpos($file, DIRECTORY_SEPARATOR);
-		if ($dirSeparator !== FALSE) {
-			$testDirsToGo[substr($file, 0, $dirSeparator)][] = $file;
+		$contents = file_get_contents($file);
+		if (preg_match('/--CONFLICTS--(.+?)--/', $contents, $matches)) {
```

Comment:
The dot won't match newlines without the `s` modifier.
Also, according to
```php
		// Match the beginning of a section.
		if (preg_match('/^--([_A-Z]+)--/', $line, $r)) {
			$section = (string)$r[1];
```
below, a section should start a line, so maybe `'/^--CONFLICTS--(.+?)^--/ms'` (keeping simple)?

Reply:
Does `/m` allow matches across multiple lines? You're definitely right that this is missing an `/s` modifier though.

Reply:
`/m` is required for `^` to match the start of a line inside the string. Alternatively you could do maybe `'/(?<=\n)--CONFLICTS--(.+?)(?<=\n)--/s'`, or just `'/\R--CONFLICTS--(.+?)\R--/s'`, but would it be better?

Reply:
Done.